### ISSUE TITLE
How to merge `dev`/`v0.41`/`v1.11` branches to `main`

### ISSUE THREAD

**Issue by huan** [wechaty/puppet-whatsapp#118]
Recently we have been heavily developing on the `dev` branch and have lots of code added.

Before we can merge the code from the `dev` branch to `main` (so that we can publish the official NPM module), I'd like to suggest the following steps as the requirement, to make the code base clearer and more readable, to make sure it can be easy continuing developed by the future maintainers.

## How to merge code from `dev` to `main`

1. create one issue for each new feature, with a concise title and a detailed description, and link to this issue. (if the feature can be split into multiple smaller ones, then it must be split into multiple smaller ones, and each one should be a separate issue)
2. create a new branch and:
    1. add your code of the implementation
    2. create unit tests to make sure your code is working as expected
    3. add the changelog to README 
4. submit a PR and link to the issue, and make sure the CI satisfy the merge requirements
5.  Wait for @windmemory or @lijiarui or @huan to merge the PR to the `main` branch, or follow their reviews and continue improving the PR.

The above list is required by the Wechaty community and we should follow those rules, to help us maintain the codebase better.

Please feel free to let me know if you have any questions/suggestions by commenting on this issue.

Thank you very much.

**Comment by windmemory** [comment]
It would be great that we merge the current `dev` or `wechaty-puppet-0.41.9` to `main` to publish a more stable version of `puppet-whatsapp`, and this will be a big milestone for our `puppet-whatsapp`.

However the requirements listed above is too strict, I have two cocerns below:
1. there are a lot of code changes and commits in the branches mentioned above, 2000+ lines code changes and 200+ commits, the workload will be heavy to separate the changes.
1. **[more important]** I am worried about the separation process might introduces unexpected bugs to the code base. If we separate the changes into multiple pieces, there might be some changes override each other, if we can not deal with these changes carefully enough, we might end up with a different code base from the original branch, and it will be really hard to debug the problem.

## My proposal
1. after we finish the current version, we submit a PR from working branch(`dev` or `wechaty-puppet-0.41.9`) to `main`
1. we write up a changelog describing everything that we have added to the new version
1. after the new version has been merged, all the following features added to the `puppet-whatsapp` will be following the rules you describe, which is every feature/PR added to the repo will have a specific branch and related issue

Can we follow my proposal to push the new version forward? Let me know what do you think @huan

**Comment by huan** [comment]
I understand that there will be additional work when we follow my listed steps above.

Your proposal is acceptable, let's follow your proposal this time so that we can finish all the work before the Chinese New Year.

Looking forward to using the Friday BOT with our newest code base for the Whatsapp Wechaty Developers' Home, and thank you everyone for your great effort in improving the Whatsapp Puppet for the Wechaty community, appreciate it!

After we have published the first version, let's start following your proposal #3 strictly: 

> after the new version has been merged, all the following features added to the puppet-whatsapp will be following the rules you describe, which is every feature/PR added to the repo will have a specific branch and related issue

## My 2 cents

BTW, I have some comments, just FYI:

Firstly, 2,000+ lines of code with 200 commits is not that heavy for a team with 5 full-time developers, especially we have finished all code already. All additional workloads are essential to improve our collaborating skills, such as issue creation, describing the feature, modulizing the code into separate PRs, conflict solving, peer reviews, teamwork, etc.

Secondly, that's exactly what we need to improve:

> we might end up with a different codebase from the original branch, and it will be really hard to debug the problem.

Collaborating is everything, and everyone should learn about teamwork, with the word "team" definition including humans, unit tests, and bugs.

**Comment by huan** [comment]
I have renamed our working branches to follow the Wechaty community `vM.N` pattern as a version branch and updated the title of this issue.

1. wechaty-puppet-0.41.9 -> v0.41
1. v1.11.57 -> v1.11

And because we are working on v1.11 as a branch, the version of the main branch has been bumped from v1.11 to v1.13.

FYI.

Reference to:

- #119

**Comment by huan** [comment]
Please create PR to merge to main, instead creating meaningless new branches.

We should always keep each modification small and related the main branch, and merge to main at last, instead of any middle branch

**Comment by huan** [comment]
The branch has been merged successfully.

Thank you very much for all your efforts on this project, @hcfw007, @su-chang, and @windmemory  !
